### DISCUSSION TITLE
Gotchas with using parent/child relationships and component hooks

### DISCUSSION THREAD

**Discussion by piedoom** [bevyengine/bevy#14506] [Q&A]
Are there any gotchas with hooks in regard to parent/child relationships?

I have a parent entity with a `Health` component. This parent entity uses an equipment system that attaches the item as a component to a child entity, enabling multiple items of the same type to be equipped. In this case, I have multiple equipped `Armor` child entities.

<img width="360" alt="image" src="https://github.com/user-attachments/assets/a02aa135-1cb0-44d9-9490-7a40aa5dd1fa">

Here, the `Armor` entities should contribute to the `total_health` of the parent component. I was previously achieving this via custom `Events` that managed the health add/subtract, as well as managing equipped child entities (as `RemovedComponent` doesn't return component data AFAICT, which I needed in my case). However, this is a fragile approach; if I manually construct an entity with equipped children and bypass events, it would create an incorrect state. The extra hitpoints would not be added, and if removed, total health could even go negative, which should be impossible.

For this reason, it seems like hooks are a great solution to this problem.

When an `Armor` child is added, I'm trying to increase the `Health` of the parent component, and the same in the reverse upon removal.

```rs
impl Component for Armor {
    const STORAGE_TYPE: bevy::ecs::component::StorageType = StorageType::Table;
    fn register_component_hooks(hooks: &mut ComponentHooks) {
        hooks.on_add(|mut world, entity, _component_id| {
            if let (Some(parent), Some(armor)) =
                (world.get::<Parent>(entity), world.get::<Self>(entity))
            {
                let armor_health = armor.health;
                if let Some(mut health) = world.get_mut::<Health>(parent.get()) {
                    health.0 += armor_health
                }
            }
        }).on_remove(|mut world, entity, _component_id| {
            if let (Some(parent), Some(armor)) =
                (world.get::<Parent>(entity), world.get::<Self>(entity))
            {
                let armor_health = armor.health;
                if let Some(mut health) = world.get_mut::<Health>(parent.get()) {
                    health.0 -= armor_health
                }
            }
        });
    }
}
```

(If you're big into spot-the-difference pictures, you might note these are exactly the same minus the plus `on_add`.)

Strangely, it seems that the `on_remove` works just as expected, whereas the `on_add` does not function at all (It stops executing at attempting to retrieve the parent component, so I suppose it doesn't exist?). Am I getting something wrong here? Is there some system ordering or something else?

Thanks!

**Comment by andriyDev** [comment]
I believe what's happening is that with_children doesn't set the parent on spawn. It firsts constructs all the children, then adds a command to push all those children at once into the children list. Basically this seems to be an optimization. 

Perhaps in an age with hooks and observers, we should avoid the optimization...

**Reply by piedoom** [reply]
I think this may be the case. It doesn't appear there is a way to manually construct a `Parent` at the moment either, so I may have to go via `Added` for now.

**Comment by piedoom** [comment]
From some nice folks in the Discord, and anyone else running into this, you can set the parent/child relationship manually with 

```rs
cmd.entity(parent_entity).add_child(entity);
cmd.entity(entity).set_parent(parent_entity);
```

And the `on_add` hooks will work as expected.
